src/gossip/service.zig

```diff
@@ -258,12 +260,40 @@ pub const GossipService = struct {
         deinitMux(&self.failed_pull_hashes_mux);
     }
 
-    /// these threads should run forever - so if they join - somethings wrong
-    /// and we should shutdown
-    fn joinAndExit(self: *Self, handle: *std.Thread) void {
-        handle.join();
-        self.exit.store(true, .unordered);
-    }
+    pub const RunHandles = struct {
+        exit: *AtomicBool,
+        receiver_thread: std.Thread,
+        packet_verifier_thread: std.Thread,
+        message_processor_thread: std.Thread,
+        message_builder_thread: ?std.Thread,
+        responder_thread: std.Thread,
+        dumper_thread: ?std.Thread,
+
+        /// If any of the threads join, all other threads will be signalled to join.
+        pub fn joinAndExit(handles: RunHandles) void {
+            inline for (@typeInfo(RunHandles).Struct.fields, 0..) |field, i| cont: {
+                comptime if (@field(std.meta.FieldEnum(RunHandles), field.name) == .exit) {
```

Comment:
is this `.exit` equality comparison equal to an "exit" string comparison ?

Reply:
It isn't a string comparison, it's an enum comparison, which is the same as an integer comparison. To be clear though, this isn't a runtime optimisation, as it's all happening at comptime. If we use `std.mem.eql`, the eval branch quota consumption is `fields.len * (sum_of_all_field_name_lengths)`; `std.meta.FieldEnum` consumes the eval branch quota only once to generate the enum (and it only consumes `fields.len` of the quota) thanks to memoization, and then the comparison itself consumes no eval branch quota, because it's a simple enum comparison (same as comparing an integer).